Removing APMs from the legacy checkout experience

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #3524

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

This PR introduces a series of changes related to the sunsetting of APMs on the legacy checkout experience (by Oct, 31st 2024):
- Slightly changes the APMs sunsetting promotional surface where it says "Stripe will end support..." to "Stripe ended support..."
- Blocks the enablement of APMs when the legacy checkout is enabled, and adds a notice above them about the deprecation
- Removes them from the checkout page

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

- Checkout and build this branch on your development environment (`tweak/remove-apms-legacy-checkout`)
- Hardcode the return value as `true` for both `are_apms_deprecated` (PHP) and `areAPMsDeprecated` (JS) functions
- Enable the legacy checkout experience

### Settings page

- As a merchant, confirm that you can see APMs disabled with a new pill on top of it explaining the deprecation on the settings page (`tweak/remove-apms-legacy-checkout`)
![Screenshot 2024-10-18 at 17 13 43](https://github.com/user-attachments/assets/c90faa45-1ed2-43b0-9f4b-7f7428a11bb1)

### Checkout page

- As a merchant, create a shortcode checkout page
- As a shopper, add a product to your cart
- Go to the shortcode checkout page
- Confirm that you cannot see any APMs
- You can even enable all of them on the `develop` branch and go back to this one and see if they are still unavailable
![Screenshot 2024-10-18 at 17 08 22](https://github.com/user-attachments/assets/4de381f9-4775-4b31-8eda-dffe0f88084d)

### Promotional surface

- As a merchant, confirm that you can see the updated copy:
![Screenshot 2024-10-18 at 17 12 09](https://github.com/user-attachments/assets/5396a760-69b2-46b1-959e-36d7f6f163b1)

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Removing APMs from the legacy checkout experience

wjrosa committed:
Adding helper function for reuse

wjrosa committed:
Simplifying helper function

wjrosa committed:
Backend changes to disable APMs

wjrosa committed:
Helper method for the backend implementation

wjrosa committed:
Fix helper method logic

wjrosa committed:
Merge branch 'develop' into tweak/remove-apms-legacy-checkout

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Blocking enablement of APMs and adding a specific deprecation pill

wjrosa committed:
Updating copy for the APMs sunsetting promotional banner

annemirasol commented on client/utils/index.js:
```diff
@@ -8,3 +8,13 @@ export const getAddToCartVariationParams = ( key ) => {
 
 	return wcAddToCartVariationParams[ key ];
 };
+
+/**
+ * Check if APMs are deprecated. Meaning it is past Oct 31st, 2024 and the legacy checkout is enabled.
+ *
+ * @return {boolean} Whether APMs are deprecated.
+ */
+export const areAPMsDeprecated = () =>
+	new Date() > new Date( '2024-10-31' ) &&
+	// eslint-disable-next-line camelcase
+	wc_stripe_settings_params.is_upe_checkout_enabled !== '1';
```
What do you think about having a single place for this logic and just passing a `are_apms_deprecated` param from the backend in `wc_stripe_settings_params`?

wjrosa replied:
Yeah, that's smarter! Done ✅

annemirasol commented:
> As a merchant, confirm that you can see APMs disabled with a new pill on top of it explaining the deprecation on the settings page (tweak/remove-apms-legacy-checkout)

- So I changed the cutoff date in the PHP and JS files (to `2024-01-31`), and turned on legacy checkout. This is what I see:

<img width="764" alt="Screenshot 2024-10-21 at 3 31 20 PM" src="https://github.com/user-attachments/assets/732161c2-08aa-41d5-8c14-ed1a4e0ea794">

---

- With the original cutoff date (`2024-10-31`) and legacy checkout still enabled, I see this:
<img width="764" alt="Screenshot 2024-10-21 at 3 33 26 PM" src="https://github.com/user-attachments/assets/d015fa69-0ed5-47c3-92b7-8f962e7528a7">

diegocurbelo commented on includes/class-wc-stripe-helper.php:
```diff
@@ -1601,4 +1605,13 @@ public static function get_klarna_preferred_locale( $store_locale, $billing_coun
+	 * @return bool Whether the APMs are deprecated.
+	 */
+	public static function are_apms_deprecated() {
+		return ( new \DateTime() )->format( 'Y-m-d' ) > '2024-10-31' && ! WC_Stripe_Feature_Flags::is_upe_checkout_enabled();
```
It should be Oct 29th (the same date mentioned in banner message)
```suggestion
		return ( new \DateTime() )->format( 'Y-m-d' ) > '2024-10-29' && ! WC_Stripe_Feature_Flags::is_upe_checkout_enabled();
```

diegocurbelo commented on client/utils/index.js:
```diff
@@ -8,3 +8,13 @@ export const getAddToCartVariationParams = ( key ) => {
+ * @return {boolean} Whether APMs are deprecated.
+ */
+export const areAPMsDeprecated = () =>
+	new Date() > new Date( '2024-10-31' ) &&
```
It should be Oct 29th (the same date mentioned in banner message)
```suggestion
	new Date() > new Date( '2024-10-29' ) &&
```

I like @annemirasol suggestion to rely on the server for this flag; that way, we don't need to worry about the customer's browser date being wrong 🤷🏼

diegocurbelo committed:
Merge branch 'develop' into tweak/remove-apms-legacy-checkout

wjrosa committed:
Update includes/class-wc-stripe-helper.php

Co-authored-by: Diego Curbelo <diego@curbelo.com>

wjrosa committed:
Update client/utils/index.js

Co-authored-by: Diego Curbelo <diego@curbelo.com>

wjrosa committed:
Merge branch 'develop' into tweak/remove-apms-legacy-checkout

wjrosa committed:
Moving JS feature flag to the JS params coming from the backend

wjrosa committed:
Fix fatal errors

wjrosa committed:
Fix JS unit tests

wjrosa committed:
Update date in comments

wjrosa committed:
Fix methods not being listed

wjrosa committed:
Update date when listing methods in settings

wjrosa committed:
Disabling customize button when method is deprecated

wjrosa commented:
> > As a merchant, confirm that you can see APMs disabled with a new pill on top of it explaining the deprecation on the settings page (tweak/remove-apms-legacy-checkout)
> 
> * So I changed the cutoff date in the PHP and JS files (to `2024-01-31`), and turned on legacy checkout. This is what I see:
> 
> <img alt="Screenshot 2024-10-21 at 3 31 20 PM" width="764" src="https://private-user-images.githubusercontent.com/730823/378588965-732161c2-08aa-41d5-8c14-ed1a4e0ea794.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MDY5NTUsIm5iZiI6MTcyOTYwNjY1NSwicGF0aCI6Ii83MzA4MjMvMzc4NTg4OTY1LTczMjE2MWMyLTA4YWEtNDFkNS04YzE0LWVkMWE0ZTBlYTc5NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyMlQxNDE3MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZDg1OWFmZmU5NjM1NWExM2E2NDdmMmJkNWU3N2NiZDgwMDdjYTEyNGM4MTlhZTVmMGYzOGZiOTMxOTZhODllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.oe4-ZUtQ25cWyMqHkAlXz2w-hYJW7hYFDk-jrwJXtVs">
> * With the original cutoff date (`2024-10-31`) and legacy checkout still enabled, I see this:
> 
> <img alt="Screenshot 2024-10-21 at 3 33 26 PM" width="764" src="https://private-user-images.githubusercontent.com/730823/378589538-d015fa69-0ed5-47c3-92b7-8f962e7528a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MDY5NTUsIm5iZiI6MTcyOTYwNjY1NSwicGF0aCI6Ii83MzA4MjMvMzc4NTg5NTM4LWQwMTVmYTY5LTBlZDUtNDdjMy05MmI3LThmOTYyZTc1MjhhNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyMlQxNDE3MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00N2JhMmYwMjgxZjliOGE3ZmZmY2RiNmJhYzE3Mjg3OTgyMDgyYTg3NjM3YjlmMjAzNmI3NzA2Nzk3YTNkZjNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.rksOieolxGoMX_O0c3hvLS4P5ShPdOTD3CNCdocuuw0">

Thanks for the review, Anne! I have fixed the issue with methods not being listed. And updated the JS code to use the backend flag instead of duplicating the logic there.

annemirasol commented on changelog.txt:
```diff
@@ -1,6 +1,7 @@
 *** Changelog ***
 
 = 8.9.0 - xxxx-xx-xx =
+* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 31st, 2024.
```
```suggestion
* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 29, 2024.
```

annemirasol commented on client/components/payment-method-deprecation-pill/index.js:
```diff
@@ -0,0 +1,56 @@
+				<Popover
+					BaseComponent={ IconComponent }
+					content={ __(
+						'This payment method is deprecated on the legacy checkout as of Oct, 29th 2024.'
```
```suggestion
						'This payment method is deprecated on the legacy checkout as of Oct 29, 2024.'
```

annemirasol commented on readme.txt:
```diff
@@ -129,6 +129,7 @@ If you get stuck, you can ask for help in the Plugin Forum.
 == Changelog ==
 
 = 8.9.0 - xxxx-xx-xx =
+* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 31st, 2024.
```
```suggestion
* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 29, 2024.
```

annemirasol commented:
@wjrosa Looks great ✨ Thanks for working on this!

I spotted some typo regarding the new date and added suggested fixes.

Minor fix requested 🙏: When a notice box/modal is open, I expect that clicking anywhere outside the box should dismiss it. Other notices, like "Requires currency", behaves that way. "Deprecated" doesn't. Worth checking if it's an easy thing to add?

<img width="300" alt="Screenshot 2024-10-22 at 10 19 59 AM" src="https://github.com/user-attachments/assets/41b53db5-1ca1-40fb-a13d-f0bffd161abf">

wjrosa committed:
Update changelog.txt

Co-authored-by: Anne Mirasol <anne.mirasol@automattic.com>

wjrosa committed:
Update client/components/payment-method-deprecation-pill/index.js

Co-authored-by: Anne Mirasol <anne.mirasol@automattic.com>

wjrosa committed:
Update readme.txt

Co-authored-by: Anne Mirasol <anne.mirasol@automattic.com>

wjrosa committed:
Fix popover dismiss when clicking outside the container

wjrosa committed:
Merge branch 'tweak/remove-apms-legacy-checkout' of https://github.com/woocommerce/woocommerce-gateway-stripe into tweak/remove-apms-legacy-checkout

wjrosa commented:
> @wjrosa Looks great ✨ Thanks for working on this!
> 
> I spotted some typo regarding the new date and added suggested fixes.
> 
> Minor fix requested 🙏: When a notice box/modal is open, I expect that clicking anywhere outside the box should dismiss it. Other notices, like "Requires currency", behaves that way. "Deprecated" doesn't. Worth checking if it's an easy thing to add?
> 
> <img alt="Screenshot 2024-10-22 at 10 19 59 AM" width="300" src="https://private-user-images.githubusercontent.com/730823/378905694-41b53db5-1ca1-40fb-a13d-f0bffd161abf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2Mjg4MzksIm5iZiI6MTcyOTYyODUzOSwicGF0aCI6Ii83MzA4MjMvMzc4OTA1Njk0LTQxYjUzZGI1LTFjYTEtNDBmYi1hMTNkLWYwYmZmZDE2MWFiZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyMlQyMDIyMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YTQzNTk4MzE2OWQ5YjY5M2MyNjhkZWE2ZWM3ZTNmZTkzMjhiMjU3ZjBkYmVkNTQ1NGE1ZjJmN2EwYjcxMTU3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.q1x_z2Vr6-gZJRHRsVaDKRdwtNzGw_JmRl5HBsDusG8">

Thanks for the suggestions! I've applied all of them. As for the popover bug, I fixed it by including a link with the Stripe page on it. The bug actually exists on the component itself (it works only when a link is present), but I thought it could be a nice solution.

annemirasol commented on client/components/payment-method-deprecation-pill/index.js:
```diff
@@ -0,0 +1,79 @@
+					mixedString:
+						/* translators: $1: a payment method name. %2: Currency(ies). */
+						__(
+							'This payment method is deprecated on the {{currencySettingsLink}}legacy checkout as of Oct, 29th 2024{{/currencySettingsLink}}.',
```
Nit: Shouldn't the comma be after `29th`?

wjrosa replied:
Oops, fixed it now!

annemirasol reviewed (approved):
Thanks for fixing the modal dismiss issue! 🚢 

✅ APMs are deprecated if after the cutoff date and legacy checkout is enabled.
✅ APMs are not deprecated if after the cutoff date and legacy checkout is enabled.
✅ APMs are not deprecated if after the cutoff date and legacy checkout is not enabled.
✅ APMs are not deprecated if before the cutoff date.

wjrosa committed:
Fix deprecation date on the new popover

wjrosa committed:
Merge branch 'develop' into tweak/remove-apms-legacy-checkout
